tests/realtikvtest/txntest/txn_test.go

```diff
@@ -532,3 +534,63 @@ func TestCheckTxnStatusOnOptimisticTxnBreakConsistency(t *testing.T) {
 	tk2.MustExec("admin check table t2")
 	tk2.MustQuery("select * from t2 order by id").Check(testkit.Rows("1 10", "2 11"))
 }
+
+func TestDMLWithAddForeignKey(t *testing.T) {
+	defer config.RestoreFunc()()
+	config.UpdateGlobal(func(conf *config.Config) {
+		conf.TiKVClient.AsyncCommit.SafeWindow = 10 * time.Second
+		conf.TiKVClient.AsyncCommit.AllowedClockDrift = 500 * time.Millisecond
+	})
+
+	store := realtikvtest.CreateMockStoreAndSetup(t)
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("set global tidb_enable_1pc='OFF';")
+	tk.MustExec("set global tidb_enable_metadata_lock='OFF';")
+	tk.MustExec("set global tidb_enable_async_commit='ON'")
+
+	tkDML := testkit.NewTestKit(t, store)
+	tkDML.MustExec("use test")
+
+	tkDDL := testkit.NewTestKit(t, store)
+	tkDDL.MustExec("use test")
+	tkDDL.MustExec("create table parent (id int primary key, val int, index(val));")
+	tkDDL.MustExec("create table child (id int primary key, val int, index(val));")
+
+	// The fail path of this test is:
+	// tk:     INSERT -> ... -> sleep(1000)                                -> PreWrite -> ... -> Async Commit -> sleep(50000) -> ... -> Success.
+	// tkDDL:            DDL -> StateWriteOnly -> checkForeignKeyConstrain -> sleep(3000)                     -> Success
+	// After fixing, either the `tkDDL` or `tk` will fail.
+	failpoint.Enable("tikvclient/beforePrewrite", "sleep(1000)")
+	defer failpoint.Disable("tikvclient/beforePrewrite")
+	failpoint.Enable("github.com/pingcap/tidb/pkg/ddl/afterCheckForeignKeyConstrain", "sleep(3000)")
+	defer failpoint.Disable("github.com/pingcap/tidb/pkg/ddl/afterCheckForeignKeyConstrain")
+	failpoint.Enable("tikvclient/asyncCommitDoNothing", "sleep(15000)")
+	defer failpoint.Disable("tikvclient/asyncCommitDoNothing")
```

Comment:
Would it be possible to use `failpoint.InjectCall` or `failpoint.WithHook` and use channels or something else to syncronize the test, instead of using sleep? Just so the test could be run a bit faster.

Reply:
Good. I've replaced the failpoints with `"pause"` and `InjectCall`. It's now faster and stabler.

Previously I used `"sleep"` because the repo `client-go` uses` failpoint.Eval` but not `failpoint.Call` to trigger the failpoint, so we cannot add hook for these failpoints :thinking:.
But now, I realize that using `"pause"` is enough to implement this test.